server/boot/customRoutes.js

```diff
@@ -4,75 +4,105 @@ module.exports = function (app) {
   const { Booking, Slot, Store } = app.models; 
   const bodyParser = require("body-parser");
   const moment = require("moment");
+  const passportSetup = require("../passport-setup");
+  const passport = require("passport");
+  const cookieSession = require("cookie-session");
   const generateTimeslots = require('./helpers').generateTimeslots; 
-  // require("mongodb-moment")(moment);
+  const keys = require("../keys");
+  const cron = require("node-cron")
+
+  app.use(cookieSession({
+    //maxAge for session 1 day 
+    maxAge: 24 * 60 * 60 * 1000,
+    keys: [keys.session.cookieKey]
+  }));
+  
+  // Initialize passport
+  app.use(passport.initialize());
+  app.use(passport.session());
+
   router.use(bodyParser.json({ extended: true }));
 
+  
+  //Cron job-- triggers each day at 1am GMT to generate the next day's slots for each store.
+  cron.schedule("0 0 * * *", function() {
```

Comment:
this might be a dumb question, but how would I call/require it here or in main? I considered putting the cron scheduler in a separate file in a fn like initCron() and then calling it from boot/main, but then the enclosing fn would only run once when called. Are you suggesting something like put the query logic in a separate fn in another file, and then use cron.schedule(cron expression, fnFromOtherFile()) in customRoutes?

This might end up being a moot point since I think we're going to switch over to dynamoDB like you suggested, and then use an AWS lambda function to trigger the cron job. (Or Antanas suggested possibly heroku cloud db). But still curious to know the answer.

Reply:
There are no dumb questions, especially since my main value add right now is answering questions :). I think your question is really about how cron jobs are queued and managed by the event loop. When the event loop encounters a setInterval/cron job, it enqueues the task as well as the callback to be executed and will continue to execute the callback at the given interval even if there are no other server tasks.  Read more about it here: https://nodejs.org/fa/docs/guides/event-loop-timers-and-nexttick/

Here are some examples that I think will clarify how cron jobs are managed; try them out and see: 

// // Example 1 - everything in the main server file (index.js)
// // to run: node index.js
const express = require('express');
const cron = require('node-cron');
app = express();
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});
app.listen(3128);

-------------------------------

// // Example 2 - main server file requires cron-job.js 
// // to run: node index.js
// server file: 
const express = require('express');
const cron = require('node-cron');
app = express();
require('./cron-job.js');
app.listen(3128);

// cron-job.js
const cron = require('node-cron');
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});

-------------------------------------

// // Example 3 - running the cron job without running the server file 
// // to run: node cron-job.js
// cron-job.js
const cron = require('node-cron');
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});

--------------------------------------

So I was suggesting that we put the cron job in a separate file as demonstrated in Example 2.

Reply:
I'm all for moving toward DynamoDb and Lambda, which is perfect for handling cron jobs. And DynamoDb will simplify database management and development, since we won't have to manange local databases. Too, these are great technologies to know before you enter the job market. Let me know if you want to discuss implementation.

Reply:
Thanks for the depth of the answer. For example 2, how come we need to require node-cron in both server.js and cron-job.js? Wouldn't it be sufficient to only require cron-job.js in server, since it already contains node-cron? 

Re: talking implementation, that would be great. I haven’t looked deeply into the implementation details yet, but I’ll do so soon. Are there some implementation decisions you’d suggest we start thinking about beforehand?

Reply:
In example 2, I left `node-cron` in server.js accidentally, you're right that it's not necessary - only cron-job.js needs `node-cron`

Reply:
@christenmartin The main thing is to understand how DynamoDB handles partition keys and sort keys. Basically, you can use a single DynamoDB table as a database, because records are identified by partition key first, then sort key. 
https://aws.amazon.com/blogs/database/choosing-the-right-dynamodb-partition-key/